### PR TITLE
[FrameworkBundle] Introduce autowirable ControllerTrait

### PR DESCRIPTION
**Pull request by dunglas**
| Q | A |
| --- | --- |
| Branch | master |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #16863 |
| License | MIT |
| Doc PR | todo |

This is the missing part of the new controller system and it's fully BC with the old one.

Used together with the existing autowiring system, #17608 and [DunglasActionBundle](https://github.com/dunglas/DunglasActionBundle) it permits to inject explicit dependencies in controllers with 0 line of config. It's a great DX improvement for Symfony.
It also has a lot of another advantages including enabling to reuse controller accros frameworks and make them easier to test. See https://dunglas.fr/2016/01/dunglasactionbundle-symfony-controllers-redesigned/ for all arguments.

Magic methods of old controllers are still available if you use this new trait in actions.

For instance, the [`BlogController::newAction`](https://github.com/symfony/symfony-demo/blob/master/src/AppBundle/Controller/Admin/BlogController.php#L70) form the `symfony-demo` can now looks like:

``` php
namespace AppBundle\Action\Admin;

use AppBundle\Form\PostType;
use AppBundle\Utils\Slugger;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Symfony\Bundle\FrameworkBundle\Controller\ControllerTrait;
use Symfony\Component\HttpFoundation\Request;
use Symfony\Component\Form\Extension\Core\Type\SubmitType;

class NewAction {
    use ControllerTrait;

    private $slugger;

    public function __construct(Slugger $slugger)
    {
        $this->slugger = $slugger;
    }

    /**
     * @Route("/new", name="admin_post_new")
     */
    public function __invoke(Request $request)
    {
        $post = new Post();
        $post->setAuthorEmail($this->getUser()->getEmail());

        $form = $this->createForm(PostType::class, $post)->add('saveAndCreateNew', SubmitType::class);
        $form->handleRequest($request);

        if ($form->isSubmitted() && $form->isValid()) {
            $post->setSlug($this->slugger->slugify($post->getTitle()));
            $entityManager = $this->getDoctrine()->getManager();
            $entityManager->persist($post);
            $entityManager->flush();

            $this->addFlash('success', 'post.created_successfully');
            if ($form->get('saveAndCreateNew')->isClicked()) {
                return $this->redirectToRoute('admin_post_new');
            }

            return $this->redirectToRoute('admin_post_index');
        }

        return $this->render('admin/blog/new.html.twig', array(
            'post' => $post,
            'form' => $form->createView(),
        ));
    }
}
```

As you can see, there is no need to register the `slugger` service in `services.yml` anymore and the dependency is explicitly injected. In fact the container is not injected in controllers anymore.

Convenience methods still work if the `ControllerTrait` is used (of course it's not mandatory). Here I've made the choice to use an invokable class but this is 100% optional, a class can still contain several actions if wanted.

Annotations like `@Route` still work too. The old `abstract` controller isn't deprecated. There is no valid reason to deprecate it IMO. People liking using the "old" way still can.

Unless in #16863, there is only one trait. This trait/class is basically a bunch of proxy methods to help newcomers. If you want to use only some methods, or want explicit dependencies (better), just inject the service you need in the constructor and don't use the trait.

I'll create open a PR on the standard edition soon to include ActionBundle and provide a dev version of the standard edition to be able to play with this new system.

I'll also backport tests added to the old controller test in 2.3+.

**Edit:** It now uses getter injection to benefit from lazy service loading by default.

### CODE REVIEW

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 1]
```diff
@@ -0,0 +1,388 @@
...
+        denyAccessUnlessGranted as public;
+        renderView as public;
+        render as public;
+        stream as public;
+        createNotFoundException as public;
+        createAccessDeniedException as public;
+        createForm as public;
+        createFormBuilder as public;
+        getDoctrine as public;
+        isCsrfTokenValid as public;
```
Comment (iltar) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Would it be an idea (while at it), to split them?
 - ControllerSessionTrait
 - ControllerSecurityTrait
 - ControllerResponseTrait
 - ControllerFormTrait
 - ControllerRouterTrait
 - ControllerDoctrineTrait
 - Missed one?

Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: As explained in the description, I think this is useless. They are proxy methods to the service itself. If you don't want the convenience of this "all-in-one" trait, just inject the service and use it.

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 2]
```diff
@@ -0,0 +1,657 @@
...
+    /**
+     * Forwards the request to another controller.
+     *
+     * @param string $controller The controller name (a string like BlogBundle:Post:index)
+     * @param array  $path       An array of path parameters
+     * @param array  $query      An array of query parameters
+     *
+     * @return Response A Response instance
+     */
+    protected function forward($controller, array $path = array(), array $query = array())
```
Comment (HeahDude) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: So I'll need to rebase #17589, when this one gets merged.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 3]
```diff
@@ -0,0 +1,388 @@
...
+        $controller = new TestTrait();
+        $controller->setAuthorizationChecker($authorizationChecker);
+
+        $this->assertTrue($controller->isGranted('foo'));
+    }
+
+    /**
+     * @expectedException \Symfony\Component\Security\Core\Exception\AccessDeniedException
+     */
+    public function testdenyAccessUnlessGranted()
```
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: `testDenyAccessUnlessGranted`

**Comment by dunglas** [comment]
Tests will be green when #18206 will be merged.

**Comment by derrabus** [comment]
@dunglas PR #16863 does not deprecate the `Controller` class either.

**Comment by dunglas** [comment]
@derrabus right, sorry about that I edit the description.

**Comment by derrabus** [comment]
@iltar I'd like to see the trait to be split as well (see #16863), to make those helpers reusable in non-controller classes like event listeners for instance. If this PR gets merged, I will rebase my PR onto it.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 4]
```diff
@@ -0,0 +1,388 @@
...
+        $controller->setHttpKernel($kernel);
+
+        $response = $controller->forward('a_controller');
+        $this->assertEquals('xml--fr', $response->getContent());
+    }
+
+    public function testRedirect()
+    {
+        $controller = new TestTrait();
+        $response = $controller->redirect('http://dunglas.fr', 301);
```
Comment (derrabus) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: I'd prefer neutral URLs in test cases, like `http://example.com` for instance.

Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: I agree. We should make this a requirement. And we could normalize person names too. Although they are tests and not "real code", it would be great to be consistent.

**Comment by derrabus** [comment]
The trait introduces some logic to fetch the depedency from the container if it hasn't been set via a setter. I don't see this logic being covered by the tests.

**Comment by dunglas** [comment]
@derrabus I've added such tests in separate PRs: #18206, #18204, #18203

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php (hunk 4 above)
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Why? It's just tests (never displayed to the end user), they are already a lot of things like that in the code base and it's fun to see some "Bernard" or "Fabien" things in test. It "humanizes" the project, I prefer that to all those "Acme Inc" and "example.com" BS.

Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: https://twitter.com/search?q=vairelles%20symfony This what I mean by "humanize".

Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: `example.com` and `example.org` are reserved for documentation and examples: http://www.iana.org/domains/reserved  They are a global standard. They are guaranteed to work forever and it will always be safe to use them. No other domain (not even `symfony.com` itself) can guarantee that.

Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: It doesn't matter here. The URL is never accessed. Even if the domain doesn't exist the test will still be green.

**Comment by dunglas** [comment]
ping @symfony/deciders

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php (hunk 4 above)
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Even if it does not matter, using `example.com` for better consistency would be good.

**Comment by weaverryan** [comment]
:-1: Unless I'm missing something, this would require (if you're autowiring your controller) *all* of these services to be eagerly instantiated in order to render your controller. Most are common services, but not all of them (form factory, serializer, etc).

I understand why some people might find this useful. I would rather see an approach that:

A) Completely leaves the `Controller` alone. I see you made the trait look for a `$this->container` property as a fallback, but it's just weird - checking for a property *and* also falling back to the container just don't look compatible.

B) Introduce smaller traits that have some of the shortcuts of the controller. And sure, you could wrap those up into one "big" trait for convenience. But even that, I would only include the *most* common services - like `router`, `logger` - not `serializer`, `form.factory`, etc.

Cheers!

**Comment by dunglas** [comment]
The check for the container in A is basically for BC and ease the maintenance (it allows to use the trait in the abstract controller class).

B is useless. This trait is just a set of proxy method. If you want only some dependencies, just inject services you use in the container.

Closing for now.

**Comment by weaverryan** [comment]
@dunglas Yea, I know that (A) was for BC - it makes sense, but it's just a little magic, with the trait looking for a property that's named `$container`.

About (B), it doesn't look useless to me. It brings 2 advantages: (A) RAD (using a trait is even less work than adding a constructor arg, though not a *ton* less work) and (B) familiarity/consistency: people can use the same shortcut methods - like `generateUrl()` - that they use in normal controllers.

**Comment by dunglas** [comment]
For A, it allows us to maintain only one trait, and it works without autowiring with any class having a `$container` property (the convention in Symfony). IMO it's the best solution on the long term to be compatible both with the legacy and enable new usages. Add a new method to the trait (like the `json` one in 3.1), and it will be also available in the traditional `abstract` controller. The same apply for bug fixes and so on.

For B, IMO there are 2 kind of users:

1. Newcomers and developers using Symfony as a RAD tool (my case, most of the time): they don't care about performance optimization and will always include the `ControllerTrait` because it's the easiest and quickest to use (and it works exactly like the good old `abstract` controller)
2. Advanced developers who care about performance, code quality and maintainability (my case on some projects): they will always carefully choose which dependencies to inject in their controllers and will explicitly inject them in the constructor (autowiring helps here, but traits are useless for them).

For instance, take a look at the `RenderHelperTrait` created by @derrabus in #16863: https://github.com/derrabus/symfony/blob/master-controller-trait/src/Symfony/Bundle/FrameworkBundle/Templating/RenderHelperTrait.php

It does almost nothing except choosing between Twig and the old templating system. I don't get the point of importing this trait instead of injecting directly an instance of `\Twig_Environment` in the constructor.

However I do get the interest of using this `ControllerTrait` because it imports automatically all common dependencies of a controller (thanks to setter autowiring).

To summarize, with this all-in-one trait I need to be aware of its existence and I can create easily controllers with all common dependencies autowired. With separated traits, I must be aware of all the existing traits and this not different than being aware of all dependencies directly (that can be autowired too).

But - I maybe missing something and I would have this merged before the feature freeze. @derrabus and you find a benefit to separated traits. What can we do to make you reconsider your :-1:? Do you want I that I split this all-in-one trait in separate traits like in #16863?

**Comment by rvanlaak** [comment]
From a developer perspective I'm not really sure whether having multiple traits as in #16863 or only having one trait is preferable. Having multiple traits looks preferable to me if your controller/service/action class would only need one trait (for instance to follow [ADR](http://pmjones.io/adr/)).

In addition, the traits are great for testing because you can [change method visibility](http://php.net/manual/en/language.oop5.traits.php#language.oop5.traits.visibility). If developers start working with Symfony, the abstract controller probably is one of the first things they will see while navigating through the source code. So in my opinion it should reflect the best practice, is there a specific one related to using traits? :+1:

**Comment by derrabus** [comment]
First of all, the `Controller` class does not have a single resposibility. It's a collection of glue code, a façade to all kinds of framework functionality. This makes it easy to use for unexperienced developers, but also allows rapid prototyping. It is also a good compromise, if you need to code a small app that will be in use for a couple of weeks/months and scrapped afterwards. Both are problems many agencies have, for instance.

Using the `Controller` class is not wrong per se. But I think, we both agree, that you should not use it if high maintainability and framework decoupling are your goals.

`Controller` is not and never will be a class with a fixed set of dependencies. This is why the `Controller` class opts-out of the dependency injection pattern pulls its dependencies on demand. This way, more of this glue code may be added in a later Symfony release with very low impact on existing applications.

I still think that autowiring the `Controller` class is a very bad idea. Whether you do this via a trait or directly does not matter to me. Let me explain why.

By eagerly injecting all possible dependencies, you will inevitably inject services that won't be used by the controller.
* Some of these dependencies will already be present (like `router` for instance), so it won't hurt to inject them.
* Others might not be present at all because the corresponding bundle is missing (Security, Monolog, Doctrine) or the specific feature has been disabled (`templating`, `serializer`). This will work out as long as the autowiring does not fail on missing dependencies.
* Others need to be instantiated first which might have a performance impact. See sensiolabs/SensioFrameworkExtraBundle#369 for such a case. Any later version of the framework might add additional dependencies, so each new framework release might slow down existing controllers even more.

Until now, performance was not an issue with the `Controller` class. But by autowiring all kinds of services instead of just the container, you're introducing performance issues. This is why I already said in my PR that I think autowiring won't be helpful for a `ControllerTrait`.

So now this PR does an interesting trick. The trait is theoretically autowirable, but this autowiring is not actually done yet because the `Controller` class still receives the container and the trait still has a fallback to the container. So it's a kind-of-hybrid approach. This leaves all existing implementations unaffected by the change. But _if_ you autowire a class using the trait, you will gain the negative impacts mentioned above with (unless I'm missing something essential) no real benefit.

Now talking about splitting the trait by topic/dependency. Yes, you're right, those traits do almost nothing. And this is exactly the point. Traits should imho only implement glue code that I would otherwise copy&paste. As soon as it contains real logic, class composition should be preferred over a trait.

So what's the benefit of such a small trait then? First of all, reusability for other classes than controllers. A trait with templating glue code might not only be helpful in controllers, but also in view listeners like [FrameworkExtraBundle's TemplateViewListener](https://github.com/sensiolabs/SensioFrameworkExtraBundle/blob/507a15f56fa7699f6cc8c2c7de4080b19ce22546/EventListener/TemplateListener.php). In one of my projects I already use a trait that provides the very handy `redirectToRoute()` method, so I can use it in some event listeners subscribed to the `kernel.request` event or in custom authentication modules.

Getting back to controllers, setter autowiring would again start make sense to me on the small traits. I would only `use` a trait with `router` glue code in a controller class that actually calls methods of that trait. I could then implement a controller without container dependency and just by adding the trait, the router would be magically present. Not sure if I would make use of this, but I kind-of like the idea.

**Comment by derrabus** [comment]
@dunglas If you're considering to split your trait and since splitting the trait is something I've already done in #16863, maybe we should join forces before you're duplicating my effort? Adding setters to my traits and moving the container-aware magic to `Controller` or `ControllerTrait` shouldn't be too hard.

**Comment by dunglas** [comment]
@derrabus why not if there is a consensus about that. Let's see the opinion of @symfony/deciders before wasting our time.

**Comment by derrabus** [comment]
Great! I just gave it a try on the router trait: https://github.com/derrabus/symfony/commit/b4827663271686218ce946d1001e41b7d148df75

Would this work for you? If this is the way to go, I could apply this pattern to the other traits as well.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+use Symfony\Component\Security\Core\Exception\AccessDeniedException;
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+
+/**
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ */
+class ControllerTraitTest extends \PHPUnit_Framework_TestCase
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Codacy Issue found: [The class ControllerTraitTest has 26 methods. Consider refactoring ControllerTraitTest to keep number of methods under 10.](https://www.codacy.com/app/git_9/symfony/file/2331978951/issues/source?bid=3207821&fileBranchId=3207858#l44)

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 6]
```diff
@@ -0,0 +1,388 @@
...
+
+        $controller = new TestTrait();
+        $controller->setRouter($router);
+
+        $this->assertEquals('/foo', $controller->generateUrl('foo'));
+    }
+
+    public function testForward()
+    {
+        $request = Request::create('/');
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Codacy Issue found: [Avoid using static access to class '\Symfony\Component\HttpFoundation\Request' in method 'testForward'.](https://www.codacy.com/app/git_9/symfony/file/2331978951/issues/source?bid=3207821&fileBranchId=3207858#l59)

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 7]
```diff
@@ -0,0 +1,657 @@
...
+
+    /**
+     * Checks the validity of a CSRF token.
+     *
+     * @param string $id    The id used when generating the token
+     * @param string $token The actual token sent with the request that should be validated
+     *
+     * @return bool
+     */
+    protected function isCsrfTokenValid($id, $token)
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Codacy Issue found: [Avoid variables with short names like $id. Configured minimum length is 3.](https://www.codacy.com/app/git_9/symfony/file/2331978825/issues/source?bid=3207821&fileBranchId=3207858#l645)

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 8]
```diff
@@ -0,0 +1,657 @@
...
+    {
+        $this->populateTemplating();
+
+        if (null !== $this->templating) {
+            $templating = $this->templating;
+
+            $callback = function () use ($templating, $view, $parameters) {
+                $templating->stream($view, $parameters);
+            };
+        } else {
```
Comment (trousers) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Codacy Issue found: [The method stream uses an else expression. Else is never necessary and you can simplify the code to work without else.](https://www.codacy.com/app/git_9/symfony/file/2331978825/issues/source?bid=3207821&fileBranchId=3207858#l502)

**Comment by weaverryan** [comment]
With all the autowiring improvements (and the DunglasActionBundle), I'm not sure of the best path forward. Things like the autodiscovery of autowireable services (i.e. DunglasActionBundle) need to be tried in the wild and taught to beginners to see how it looks/feels.

For that reason, I don't think we should consider this PR and others like #18300 for 3.1. *If* setter injection (#17608) were merged, then this could be implemented in DunglasActionBundle, and we could try and test it for the future.

**Comment by TomasVotruba** [comment]
I would like to try this feature in practice before merging, since it is great change that may go wrong.

**So I've implemented these traits for Controllers in bundle**:
- http://www.tomasvotruba.cz/blog/2016/12/12/decouple-controller-from-symfony-using-smart-traits/
- https://github.com/Symplify/ControllerAutowire

Feel free to try it and write your experience and what was good and bad.

It might help to discuss this PR in more real-world way.

**Comment by dunglas** [comment]
Hi,

Nice! If you want to test this PR in particular, I'm also maintaining a temporary fork with some pending autowiring related things (including this PR): https://github.com/dunglas/symfony/tree/autowiring

**Comment by dunglas** [comment]
The PR now uses getter injection (#20973) to benefit form lazy service loading by default. The performance penalty of using setter injection is gone!

**Comment by dunglas** [comment]
To enable getter autowiring support in this `ControllerTrait`, we need PHP 7+ (for return types). I've reverted changes made to the `Controller` class to let it compatible with PHP 5 (it doesn't use the new trait anymore).

There is some code duplication between the trait and the controller class. It will make the maintenance a bit harder, on the other hand it removes the magic check on the `$container` property in the trait. If we want to avoid this duplication, maybe should we think to deprecate the `Controller` class and promote the new trait instead.

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 9]
```diff
@@ -0,0 +1,433 @@
...
+use Symfony\Component\Form\Extension\Core\Type\FormType;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+use Symfony\Bundle\FrameworkBundle\Templating\EngineInterface;
+
+/**
+ * Common features needed in controllers.
+ *
+ * Supports both autowiring trough setters and accessing services using a container.
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Needs update

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 10]
```diff
@@ -0,0 +1,433 @@
...
+    {
+        throw new \LogicException(sprintf('An instance of "%s" must be provided.', Session::class));
+    }
+
+    protected function getAuthorizationChecker(): AuthorizationCheckerInterface
+    {
+        throw new \LogicException(sprintf('An instance of "%s" must be provided.', AuthorizationCheckerInterface::class));
+    }
+
+    protected function getTemplating(): EngineInterface
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Should we remove this one since templating is going to be deprecated ?

**Review by nicolas-grekas — COMMENTED** [review]
About code duplication, I think we don't care.
About the deprecation of the base controller class, I think it's a good idea to do it in the same PR.

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php (hunk 10 above)
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Yes

**Comment by dunglas** [comment]
* Marked as experimental
* Templating support removed
* Rebased

Now that #21031 has been merged, this PR is ready to be merged too.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by chalasr** [comment]
A great step forward. 👍

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+use Symfony\Component\Security\Core\User\User;
+use Symfony\Component\Security\Csrf\CsrfTokenManagerInterface;
+use Symfony\Component\Serializer\SerializerInterface;
+
+/**
+ * @author Kévin Dunglas <dunglas@gmail.com>
+ *
+ * @requires PHP 7
+ */
+class ControllerTraitTest extends \PHPUnit_Framework_TestCase
```
Comment (xabbuh) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: Should probably be updated to use the namespaced `TestCase` class.

**Comment by robfrawley** [comment]
@dunglas How does one unit test any class with getter injection? Do you need to implement a class that extends the code you want to test with implementations for all the getters? Or are there other methods for this?

**Comment by nicolas-grekas** [comment]
Looks like you missed http://symfony.com/blog/new-in-symfony-3-3-getter-injection :)

**Comment by robfrawley** [comment]
@nicolas-grekas I did; thanks! :+1:

File: src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php (hunk 11 above)
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: ping @dunglas

Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Controller/ControllerTraitTest.php]: done

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php [hunk 12]
```diff
@@ -0,0 +1,443 @@
...
+ * @experimental in version 3.3
+ */
+trait ControllerTrait
+{
+    /**
+     * @required
+     */
+    protected function getRouter(): RouterInterface
+    {
+        throw new \LogicException(sprintf('An instance of "%s" must be provided.', RouterInterface::class));
```
Comment (fabpot) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: I would remove the exceptions here as PHP would throw an error anyway thanks to the type hint.

**Comment by fabpot** [comment]
@dunglas Can you also add a note in the CHANGELOG (specifying that this new feature requires PHP 7, and perhaps mentioning the few differences with the Controller base class)? Thanks.

File: src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php (hunk 12 above)
Comment (dunglas) [inline src/Symfony/Bundle/FrameworkBundle/Controller/ControllerTrait.php]: Done.

I've also removed the conditional behavior of the `json()` method to make the Serializer mandatory:
* it's a cleaner design (the behavior is predictable)
* it allows to not `catch` a `TypeError`
* having a proxy method doing `return new JsonResponse($data, $status, $headers)` is useless

**Comment by dunglas** [comment]
done

**Comment by fabpot** [comment]
👍

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by fabpot** [comment]
Thank you @dunglas.

**Comment by fabpot** [comment]
see #22157 as this was more or less reverted by that PR.
